crates/node/src/storage/rocksdb.rs

```diff
@@ -688,12 +734,148 @@ impl RocksDB {
             old_and_new_diff_key(key, height)?.1
         };
 
-        self.0
+        self.inner
             .get_cf(rollback_cf, key)
             .map_err(|e| Error::DBError(e.into_string()))
     }
 }
 
+pub struct DbSnapshot<'a>(pub rocksdb::Snapshot<'a>);
+
+impl<'a> DbSnapshot<'a> {
+    /// Write a snapshot of the database out to file. The last line
+    /// of the file contains metadata about how to break the file into
+    /// chunks.
+    pub fn write_to_file(
+        &self,
+        cfs: [(&'static str, &'a ColumnFamily); 6],
+        path: &Path,
+    ) -> std::io::Result<()> {
+        let file = File::create(path)?;
+        let mut buf = BufWriter::new(file);
+        let mut chunker = Chunker::new(MAX_CHUNK_SIZE);
+        for (cf_name, cf) in cfs {
+            let read_opts = make_iter_read_opts(None);
+            let iter =
+                self.0.iterator_cf_opt(cf, read_opts, IteratorMode::Start);
+
+            for (key, raw_val, _gas) in PersistentPrefixIterator(
+                PrefixIterator::new(iter, String::default()),
+                // Empty string to prevent prefix stripping, the prefix is
+                // already in the enclosed iterator
+            ) {
+                let val = HEXLOWER.encode(&raw_val);
+                let bytes = format!("{cf_name}:{key}={val}\n");
+                chunker.add_line(&bytes);
+                buf.write_all(bytes.as_bytes())?;
+            }
+            buf.flush()?;
+        }
+        let chunks = chunker.finalize();
+        let val = HEXLOWER.encode(&chunks.serialize_to_vec());
+        let bytes = format!("chunks:{val}");
+        buf.write_all(bytes.as_bytes())?;
+        buf.flush()?;
+        Ok(())
+    }
+
+    /// Remove snapshots older than the latest
+    pub fn cleanup(
+        latest_height: BlockHeight,
+        base_dir: &Path,
+    ) -> std::io::Result<()> {
+        let toml = OsStr::new("toml");
+        for entry in std::fs::read_dir(base_dir)? {
+            let entry = entry?;
+            if entry.path().is_file() && Some(toml) == entry.path().extension()
+            {
+                if let Some(name) = entry.path().file_name() {
+                    let Some(height) = name
+                        .to_string_lossy()
+                        .strip_prefix("snapshot_")
+                        .and_then(|n| n.strip_suffix(".toml"))
+                        .and_then(|h| BlockHeight::from_str(h).ok())
```

Comment:
this shit is kinda gnarly 😂

Reply:
I agree 😫 Do you have a suggestion?

Reply:
well, not really. when I write this kinda shit I typically use `split_once`, which isn't that much better. I follow it up with a but of `let-else` bindings to extract the tokens I'm after. string parsing is always going to look ugly